src/linear_algebra/char_poly/coeff.lean

```diff
@@ -0,0 +1,162 @@
+  rw ← nat.pred_eq_sub_one, apply nat.pred_lt, apply h,
+end
+
+theorem nat_degree_char_poly_eq_dim [nontrivial R] (M : matrix n n R) :
+  (char_poly M).nat_degree = fintype.card n :=
+nat_degree_eq_of_degree_eq_some (degree_char_poly_eq_dim M)
+
+lemma char_poly_monic_of_nontrivial [nontrivial R] (M : matrix n n R) :
+  monic (char_poly M) :=
+begin
+  by_cases fintype.card n = 0, rw [char_poly, det_of_card_zero h], apply monic_one,
+  have mon : (∏ (i : n), (X - C (M i i))).monic,
+  { apply monic_prod_of_monic univ (λ i : n, (X - C (M i i))), simp [monic_X_sub_C], },
+  rw ← sub_add_cancel (∏ (i : n), (X - C (M i i))) (char_poly M) at mon,
+  rw monic at *, rw leading_coeff_add_of_degree_lt at mon, rw ← mon,
+  rw degree_char_poly_eq_dim, rw ← neg_sub, rw degree_neg,
+  apply lt_trans (char_poly_sub_diagonal_degree_lt M), rw with_bot.coe_lt_coe,
+  rw ← nat.pred_eq_sub_one, apply nat.pred_lt, apply h,
+end
+
+lemma char_poly_monic (M : matrix n n R) :
+  monic (char_poly M) :=
+begin
+  classical, by_cases h : nontrivial R,
+  { letI := h, apply char_poly_monic_of_nontrivial, },
+  { rw nontrivial_iff at h, push_neg at h, apply h, }
+end
+
+theorem trace_eq_char_poly [nonempty n] (M : matrix n n R) :
+  (matrix.trace n R R) M = -(char_poly M).coeff (fintype.card n - 1) :=
+begin
+  by_cases nontrivial R; try { rw not_nontrivial_iff_subsingleton at h }; haveI := h, swap,
+  { apply subsingleton.elim },
+  rw char_poly_coeff_eq_prod_coeff_of_le, swap, refl,
+  rw [fintype.card, prod_X_sub_C_coeff_card_pred univ (λ i : n, M i i)], simp,
+  rw [← fintype.card, fintype.card_pos_iff], apply_instance,
+end
+
+-- I feel like this should use polynomial.alg_hom_eval₂_algebra_map
+lemma mat_poly_equiv_eval (M : matrix n n (polynomial R)) (r : R) (i j : n) :
+  (mat_poly_equiv M).eval ((scalar n) r) i j = (M i j).eval r :=
+begin
+  unfold polynomial.eval, unfold eval₂,
+  transitivity finsupp.sum (mat_poly_equiv M) (λ (e : ℕ) (a : matrix n n R),
+    (a * (scalar n) r ^ e) i j),
+  { unfold finsupp.sum, rw sum_apply, rw sum_apply, dsimp, refl, },
+  { simp_rw ← (scalar n).map_pow, simp_rw ← (matrix.scalar.commute _ _).eq,
+    simp only [coe_scalar, matrix.one_mul, ring_hom.id_apply,
+      smul_val, mul_eq_mul, algebra.smul_mul_assoc],
+    have h : ∀ x : ℕ, (λ (e : ℕ) (a : R), r ^ e * a) x 0 = 0 := by simp,
+    symmetry, rw ← finsupp.sum_map_range_index h, swap, refl,
+    refine congr (congr rfl _) (by {ext, rw mul_comm}), ext, rw finsupp.map_range_apply,
+    simp [apply_eq_coeff], }
+end
+
+lemma eval_det (M : matrix n n (polynomial R)) (r : R) :
+  polynomial.eval r M.det = (polynomial.eval (matrix.scalar n r) (mat_poly_equiv M)).det :=
+begin
+  rw [polynomial.eval, ← coe_eval₂_ring_hom, ring_hom.map_det],
+  apply congr_arg det, ext, symmetry, convert mat_poly_equiv_eval _ _ _ _,
+end
+
+theorem det_eq_char_poly (M : matrix n n R) :
+  M.det = (-1)^(fintype.card n) * (char_poly M).coeff 0:=
```

Comment:
```suggestion
theorem det_eq_sign_char_poly_coeff (M : matrix n n R) :
  M.det = (-1)^(fintype.card n) * (char_poly M).coeff 0 :=
```

Reply:
do you think the longer name is easier to guess than the shorter name?

Reply:
I think the short name is false. :-)

Reply:
Of course, many names taken too literally are false, but this one is just too vague. I wouldn't insist on the `sign` part, but `coeff` is essential here.

Reply:
hmm. in my mind it's not false but rather just needs the elaborator to fill in a hole, and there's a canonical thing that fits in the hole. If you think this is a bad notion of correctness for names, I'll defer to you.

Reply:
What hole? I'm confused --- there are zillion things you can do to a polynomial to get a number, not just take a coefficient. There's no way that elaboration alone can deduce you're meant to be taking a coefficient here.

Reply:
The point is it's only "canonical" if you already know the maths.

Reply:
Okay, I can get on board with a principle like "the fact that the name has a canonical interpretation should not be mathematically significant."